### PR TITLE
Classification with x-transformers

### PR DESCRIPTION
**Pull request by RyanKim17920**
Added cls token/pooling option for NLP based full text classification

### CODE REVIEW

File: tests/classification_test.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+from x_transformers import TransformerWrapper, Encoder
+import torch
+from torch import nn
+
+# CLS token test
+transformer = TransformerWrapper(
```
Comment (lucidrains) [inline tests/classification_test.py:6]: do you want to try moving all of this under a function like `def test_classification():` ?

Comment (lucidrains) [reply tests/classification_test.py:6]: you can just slap it all into `test_x_transformers.py`

File: x_transformers/x_transformers.py [hunk 2]
```diff
@@ -1913,6 +1915,15 @@ def __init__(
         self.emb_dim = emb_dim
         self.num_tokens = num_tokens
 
+        self.pooling = pooling
+        self.use_pooling = exists(pooling)
+        self.use_cls = use_cls
+        assert not (self.use_cls and self.use_pooling), 'cannot use both pooling and cls token'
+        if self.use_cls:
+            max_seq_len+=1
```
Comment (lucidrains) [inline x_transformers/x_transformers.py:1923]: instead of changing the `max_seq_len`, want to concat the cls token after adding the absolute positional embedding?

File: x_transformers/x_transformers.py [hunk 3]
```diff
@@ -2106,14 +2122,26 @@ def forward(
 
             x = x[:, :n]
 
-        # projecting to logits
+        # cls token or pooling for classification
+        if self.use_cls:
+            x = x[:, 0]
+        
+        if self.use_pooling:
+            x = self.pooling(x).squeeze()
```
Comment (lucidrains) [inline x_transformers/x_transformers.py:2130]: for the pooling, we need to account for masking (masked averaging)

File: tests/classification_test.py (hunk 1 above)
Comment (RyanKim17920) [reply tests/classification_test.py:6]: Sure, I'll do that. Additionally, I'm trying to make an example section for IMBD dataset classification + minor fixes

File: x_transformers/x_transformers.py (hunk 3 above)
Comment (lucidrains) [reply x_transformers/x_transformers.py:2130]: i can take care of this if you'd like, it is all around a bit tricky

Comment (RyanKim17920) [reply x_transformers/x_transformers.py:2130]: Yes, please do so. Thank you 👍

**Comment by lucidrains** [comment]
@RyanKim17920 do you want to try the latest changes and see if that's enough?

**Comment by lucidrains** [comment]
@RyanKim17920 hey Ryan, sorry for hijacking your efforts, just that the project is at a size where things need to be a bit more particular

your example should run now as

```python
import torch
from torch import nn

from x_transformers import (
    TransformerWrapper,
    Encoder
)

# CLS token test
transformer = TransformerWrapper(
    num_tokens=6,
    max_seq_len=10,
    logits_dim=2, # num_classes 
    use_cls_token=True,
    attn_layers = Encoder(
        dim = 6,
        depth = 1,
        heads = 2,
    )
)

x = torch.randint(0, 5, (2, 10))
y = torch.tensor([0, 1])

print(x.shape)
logits = transformer(x)
print(logits.shape)
loss = nn.CrossEntropyLoss()(logits, y)

print(loss)

# BCE cls token

transformer = TransformerWrapper(
    num_tokens=6,
    max_seq_len=10,
    logits_dim=1, # num_classes 
    use_cls_token=True,
    squeeze_out_last_dim = True,
    attn_layers = Encoder(
        dim = 6,
        depth = 1,
        heads = 2,
    )
)

x = torch.randint(0, 5, (2, 10)).float()
y = torch.tensor([0, 1]).float()

print(x.shape)
logits = transformer(x).squeeze()
loss = nn.BCEWithLogitsLoss()(logits, y)

print(loss)

# pooling test
transformer = TransformerWrapper(
    num_tokens=6,
    max_seq_len=10,
    logits_dim=2, # num_classes 
    average_pool_embed = True,
    attn_layers = Encoder(
        dim = 6,
        depth = 1,
        heads = 2,
    )
)

x = torch.randint(0, 5, (2, 10))
y = torch.tensor([0, 1])

print(x.shape)
logits = transformer(x)
print(logits.shape)
loss = nn.CrossEntropyLoss()(logits, y)

print(loss)

# pooling BCE test

# pooling test
transformer = TransformerWrapper(
    num_tokens=6,
    max_seq_len=10,
    logits_dim=1, # num_classes 
    average_pool_embed = True,
    squeeze_out_last_dim = True,
    attn_layers = Encoder(
        dim = 6,
        depth = 1,
        heads = 2,
    )
)

x = torch.randint(0, 5, (2, 10)).float()
y = torch.tensor([0, 1]).float()

print(x.shape)
logits = transformer(x).squeeze()
print(logits.shape)
loss = nn.BCEWithLogitsLoss()(logits, y)

print(loss)

# normal test 

transformer = TransformerWrapper(
    num_tokens=6,
    max_seq_len=10,
    logits_dim=2, # num_classes 
    average_pool_embed = True,
    attn_layers = Encoder(
        dim = 6,
        depth = 1,
        heads = 2,
    )
)

x = torch.randint(0, 5, (1, 10))
y = torch.tensor([0])

print(x.shape)
logits = transformer(x)
print(logits.shape)
```

**Comment by RyanKim17920** [comment]
Thank you for the improvements you've already made to my original additions. I noticed that the test/x_transformers are outdated, so those changes aren't needed anymore. However, I believe the example I provided could still be valuable. It demonstrates the usage of the NLP classification with a well-known dataset, which might be useful for users to understand how to implement it.

Would it be possible to add the example to the repository?
